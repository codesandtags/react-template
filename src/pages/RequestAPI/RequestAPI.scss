@import '../../../src/styles/main';

.RequestAPI {
  &__title {
    margin-bottom: 3rem;
  }

  &__field {
    margin: 2rem auto;
  }

  &__results {
    margin: 3rem auto;
    border: .1rem dashed $brand-light-purple;
    padding: 2rem;
  }
}

.Selector {
  &--full-width {
    width: 100%;
    max-width: 100%;
  }
}

Button {
  margin: 0.5rem auto;
}
