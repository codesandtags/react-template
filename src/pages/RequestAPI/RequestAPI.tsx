import React, { ReactNode } from 'react';

import API from './api';

import './RequestAPI.scss';

type MyProps = {
    children: ReactNode;
};
type MyState = {
    users: {
        id: number;
        name: string;
    }[];
    user: {
        name: string;
        id?: number;
    };
    endpoint: string;
};

class RequestAPI extends React.Component<MyProps, MyState> {
    constructor(props: MyProps) {
        super(props);
        this.state = {
            users: [],
            user: {
                name: '',
            },
            endpoint: 'users',
        };
    }

    async handleGet(): Promise<void> {
        const response = await API.get(this.state.endpoint);

        this.setState({
            users: response.data,
        });
    }

    public handlePost(): void {
        const body = {
            name: this.state.user.name,
        };
        const config = {
            headers: {
                'Content-type': 'application/json; charset=UTF-8',
            },
        };

        API.post(this.state.endpoint, body, config).then(response => {
            this.setState({
                user: response.data,
                users: [response.data],
            });
        });
    }

    public handlePut(): void {
        const url = `${this.state.endpoint}/1`;
        const body = {
            id: 1,
            name: 'Edwin Torres',
            username: 'Bret',
        };
        const config = {
            headers: {
                'Content-type': 'application/json; charset=UTF-8',
            },
        };

        API.put(url, body, config).then(response => {
            this.setState({
                user: response.data,
                users: [response.data],
            });
        });
    }

    public handleDelete(): void {
        const url = `${this.state.endpoint}/1`;

        API.delete(url).then(() => {
            this.setState({
                users: [],
                user: {
                    name: '',
                },
            });
        });
    }

    public handleFieldChange(event: { target: HTMLInputElement; }): void {
        this.setState({
            user: {
                name: event.target.value,
            },
        });
    }

    public renderField(): React.ReactChild {
        return (
            <div className="RequestAPI__field">
                <div className="Column__offset-md-2">
                    <input
                        id="text-field-1"
                        placeholder="e.g. Jonh Doe"
                        onChange={this.handleFieldChange.bind(this)}
                    />
                </div>
            </div>
        );
    }

    public renderButtons(): React.ReactChild {
        return (
            <div className="RequestAPI__field">
                <div>
                    <button className="RequestAPI__button-get" onClick={this.handleGet.bind(this)}>
                        GET
                    </button>
                </div>
                <div>
                    <button className="RequestAPI__button-post" onClick={this.handlePost.bind(this)}>
                        POST
                    </button>
                </div>
                <div>
                    <button className="RequestAPI__button-put" onClick={this.handlePut.bind(this)}>
                        PUT
                    </button>
                </div>
                <div>
                    <button className="RequestAPI__button-delete" onClick={this.handleDelete.bind(this)}>
                        DELETE
                    </button>
                </div>
            </div>
        );
    }

    public renderResults() {
        const userList =
            this.state.users &&
            this.state.users.map(user => {
                return <li key={user.id}>{user.name}</li>;
            });

        let results = <div></div>;

        if (userList.length > 0) {
            results = (
                <div>
                    <div className="RequestAPI__results">
                        <div>{userList}</div>
                    </div>
                </div>
            );
        }

        return results;
    }

    render(): React.ReactChild {
        return (
            <div className="RequestAPI">
                <div className="RequestAPI__title">
                    <div>
                        <p>
                            Here you will find an example using <strong>Axios</strong> in order to consume a public API.
                            This example uses the{' '}
                            <a className="Link__link" href="https://jsonplaceholder.typicode.com" target="_blank"
                               rel="noopener noreferrer">
                                jsonplaceholder
                            </a>{' '}
                            public API.
                        </p>
                    </div>
                </div>

                {this.renderField()}
                {this.renderButtons()}
                {this.renderResults()}
            </div>
        );
    }
}

export default RequestAPI;
