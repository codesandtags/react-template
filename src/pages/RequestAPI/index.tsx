import RequestAPI from './RequestAPI';

export default RequestAPI;
