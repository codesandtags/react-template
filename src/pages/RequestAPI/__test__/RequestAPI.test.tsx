import React from 'react';
import { shallow } from 'enzyme';
import MockAdapter from 'axios-mock-adapter';
import axios from '../api';

import RequestAPI from '../';

describe('RequestAPI', () => {
  let wrapper;
  const axiosMock = new MockAdapter(axios);
  const wait = time => new Promise(resolve => setTimeout(resolve, time));
  const props = {
    children: null
  };

  afterEach(() => {});

  test('should match the snapshot', () => {
    wrapper = shallow(<RequestAPI  {...props}/>);
    expect(wrapper.html()).toMatchSnapshot();
  });

  describe('When clicks the GET button', () => {
    test('should set the users from the response', async () => {
      const spy = jest.spyOn(RequestAPI.prototype, 'handleGet');
      const mockDataGet = [
        { id: 1, name: 'John Smith' },
        { id: 2, name: 'Edwin Torres' },
      ];
      axiosMock.onGet('users').reply(200, mockDataGet);

      wrapper = shallow(<RequestAPI  {...props}/>);
      wrapper.find('.RequestAPI__button-get').simulate('click');
      await wait(0);

      expect(spy).toHaveBeenCalled();
      expect(wrapper.state().users).toEqual(mockDataGet);
    });
  });

  describe('When clicks the POST button', () => {
    test('should set the users from the response', async () => {
      const spy = jest.spyOn(RequestAPI.prototype, 'handlePost');
      const mockDataPost = {
        id: 1,
        name: 'John Smith',
      };
      axiosMock.onPost('/users').reply(200, mockDataPost);

      wrapper = shallow(<RequestAPI  {...props}/>);
      wrapper.find('.RequestAPI__button-post').simulate('click');
      await wait(0);

      expect(spy).toHaveBeenCalled();
      expect(wrapper.state().users).toEqual([mockDataPost]);
    });
  });

  describe('When clicks the PUT button', () => {
    test('should set the users from the response', async () => {
      const spy = jest.spyOn(RequestAPI.prototype, 'handlePut');
      const mockDataPut = {
        id: 1,
        name: 'John Smith',
      };
      axiosMock.onPut('/users/1').reply(200, mockDataPut);

      wrapper = shallow(<RequestAPI  {...props}/>);
      wrapper.find('.RequestAPI__button-put').simulate('click');
      await wait(0);

      expect(spy).toHaveBeenCalled();
      expect(wrapper.state().users).toEqual([mockDataPut]);
    });
  });

  describe('When clicks the DELETE button', () => {
    test('should set the users from the response', async () => {
      const spy = jest.spyOn(RequestAPI.prototype, 'handleDelete');
      const mockDataDelete = {};
      axiosMock.onDelete('/users/1').reply(200, mockDataDelete);

      wrapper = shallow(<RequestAPI  {...props}/>);
      wrapper.find('.RequestAPI__button-delete').simulate('click');
      await wait(0);

      expect(spy).toHaveBeenCalled();
      expect(wrapper.state().users).toEqual([]);
    });
  });

  describe('When types something the TextField', () => {
    test('should set the value for the user name', () => {
      const spy = jest.spyOn(RequestAPI.prototype, 'handleFieldChange');
      const userExpected = {
        name: 'Hello',
      };

      wrapper = shallow(<RequestAPI  {...props}/>);
      wrapper.find('#text-field-1').simulate('change', { target: { value: 'Hello' } });

      expect(spy).toHaveBeenCalled();
      expect(wrapper.state('user')).toMatchObject(userExpected);
    });
  });
});
