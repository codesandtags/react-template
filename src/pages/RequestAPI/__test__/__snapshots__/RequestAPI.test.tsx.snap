// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RequestAPI should match the snapshot 1`] = `"<div class=\\"Grid__container RequestAPI\\"><div class=\\"Row__container RequestAPI__title\\"><div class=\\"Column__container Column__xs-4 Column__sm-8 Column__md-12 Column__lg-12\\"><p class=\\"TextBody__text--2\\"> Here you will find an example using <strong>Axios</strong> in order to consume a public API. This example uses the <a class=\\"Link__link\\" href=\\"https://jsonplaceholder.typicode.com\\" target=\\"_blank\\" rel=\\"noopener noreferrer\\">jsonplaceholder</a> public API.</p></div></div><div class=\\"Row__container RequestAPI__field\\"><div class=\\"Column__container Column__xs-4 Column__sm-8 Column__md-8 Column__lg-8 Column__offset-md-2\\"><div class=\\"InputContainer__input\\"><label class=\\"Label__label\\" for=\\"text-field-1-input\\" label=\\"Data\\" id=\\"text-field-1-label\\">Data<span class=\\"Label__label--inline Label__label--secondary\\">(Data to send)</span></label><input type=\\"text\\" id=\\"text-field-1-input\\" value=\\"\\" placeholder=\\"e.g. Jonh Doe\\" class=\\"Input__input\\"/></div></div></div><div class=\\"Row__container Row__container--justify-center RequestAPI__field\\"><div class=\\"Column__container Column__xs-4 Column__sm-4 Column__md-2 Column__lg-2\\"><button class=\\"ButtonCore__button Button__button--primary RequestAPI__button-get\\"><span class=\\"ButtonCore__block\\"><span class=\\"ButtonCore__text\\">GET</span></span></button></div><div class=\\"Column__container Column__xs-4 Column__sm-4 Column__md-2 Column__lg-2\\"><button class=\\"ButtonCore__button Button__button--secondary RequestAPI__button-post\\"><span class=\\"ButtonCore__block\\"><span class=\\"ButtonCore__text\\">POST</span></span></button></div><div class=\\"Column__container Column__xs-4 Column__sm-4 Column__md-2 Column__lg-2\\"><button class=\\"ButtonCore__button Button__button--secondary RequestAPI__button-put\\"><span class=\\"ButtonCore__block\\"><span class=\\"ButtonCore__text\\">PUT</span></span></button></div><div class=\\"Column__container Column__xs-4 Column__sm-4 Column__md-2 Column__lg-2\\"><button class=\\"ButtonCore__button Button__button--secondary RequestAPI__button-delete\\"><span class=\\"ButtonCore__block\\"><span class=\\"ButtonCore__text\\">DELETE</span></span></button></div></div><div></div></div>"`;
